rust_src/src/lisp.rs

```diff
@@ -394,6 +394,101 @@ impl LispObject {
     }
 }
 
+impl LispObject {
+    pub fn is_thread(self) -> bool {
+        self.as_vectorlike().map_or(false, |v| {
+            v.is_pseudovector(PseudovecType::PVEC_THREAD)
+        })
+    }
+
+    pub fn is_mutex(self) -> bool {
+        self.as_vectorlike().map_or(false, |v| {
+            v.is_pseudovector(PseudovecType::PVEC_MUTEX)
+        })
+    }
+
+    pub fn is_condition_variable(self) -> bool {
+        self.as_vectorlike().map_or(false, |v| {
+            v.is_pseudovector(PseudovecType::PVEC_CONDVAR)
+        })
+    }
+
+    pub fn is_byte_code_function(self) -> bool {
+        self.as_vectorlike().map_or(false, |v| {
+            v.is_pseudovector(PseudovecType::PVEC_COMPILED)
+        })
+    }
+
+    pub fn is_subr(self) -> bool {
+        self.as_vectorlike().map_or(false, |v| {
+            v.is_pseudovector(PseudovecType::PVEC_SUBR)
+        })
+    }
+
+    pub fn is_buffer(self) -> bool {
+        self.as_vectorlike().map_or(false, |v| {
+            v.is_pseudovector(PseudovecType::PVEC_BUFFER)
+        })
+    }
+
+    pub fn is_char_table(self) -> bool {
+        self.as_vectorlike().map_or(false, |v| {
+            v.is_pseudovector(PseudovecType::PVEC_CHAR_TABLE)
+        })
+    }
+
+    pub fn is_bool_vector(self) -> bool {
+        self.as_vectorlike().map_or(
+            false,
+            |v| v.is_pseudovector(PseudovecType::PVEC_BOOL_VECTOR),
+        )
+    }
+
+    pub fn is_array(self) -> bool {
+        self.is_vector() || self.is_string() || self.is_char_table() || self.is_bool_vector()
```

Comment:
This *could* be made more efficient by only matching once for `is_vectorlike`, and then ORing the subtype checks.

But it's probably not worth the slightly smaller readability, or even optimized away in release.